From 853c3d23cb481e80e160153e297c86a6bc8ad475 Mon Sep 17 00:00:00 2001
From: pch666777 <1005100717@qq.com>
Date: Sun, 13 Jul 2025 09:29:34 +0800
Subject: [PATCH] Fix MIPS asm branch out of range

---
 src/core/CMakeLists.txt | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/core/CMakeLists.txt b/src/core/CMakeLists.txt
index 86201bc0e0..5340578f2f 100644
--- a/src/core/CMakeLists.txt
+++ b/src/core/CMakeLists.txt
@@ -2196,6 +2196,10 @@ else()
     set_source_files_properties(${CMAKE_CURRENT_BINARY_DIR}/qgsexpression_texts.cpp PROPERTIES COMPILE_FLAGS "-O1")
     set_source_files_properties(proj/qgscoordinatereferencesystem.cpp PROPERTIES COMPILE_FLAGS "-mllvm -inline-threshold=128")
   endif()
+  if("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "mips")
+	message(STATUS "MIPS architecture, adding -mlong-calls")
+	set_source_files_properties(textrenderer/qgsfontmanager.cpp PROPERTIES COMPILE_FLAGS "-mlong-calls")
+  endif()
 endif()
 
 # Generate cpp+header file from .proto file using "protoc" tool (to support MVT encoding of vector tiles)
-- 
2.50.1

